### PR TITLE
(WIP) Match padding calculations to keras

### PR DESCRIPTION
**Pull request by thefiddler**
Update padding calculations to match keras. Keras bases its padding calculations on the kernel size and dilation (if any), not on the shape of the input tensor. This fixes the padding mode="same" for dilated convolutions (issue #307).

By following the same approach, we can now also convert models with undefined input shapes, i.e. `(3, None, None)` rather than `(3, 224, 224)`.

### CODE REVIEW

**Comment by Huyuwei** [comment]
As far as I know, padding calculations in keras do depend on the input size, at least when using tensorflow backend https://www.tensorflow.org/versions/r0.12/api_docs/python/nn/convolution

Now 'same' padding for dilated convolutions has a bug as reported, and this should be fixed by
```python
kernel_h = (kernel_h - 1) * dilation[0] + 1
kernel_w = (kernel_w - 1) * dilation[1] + 1
```

Even if we convert it correctly on the frontend, now only dilation==1 is supported by TVM on the backend.

**Comment by thefiddler** [comment]
According to the [v1.4 tensorflow documentation](https://www.tensorflow.org/api_docs/python/tf/nn/convolution), the padding only depends on the kernel size and dilation, not the input shape:
If padding == "SAME":
```python
output_spatial_shape[i] = ceil(input_spatial_shape[i] / strides[i])
```

If padding == "VALID":
```python
output_spatial_shape[i] = ceil((input_spatial_shape[i] - (spatial_filter_shape[i]-1) * dilation_rate[i]) / strides[i]).
```
(Of course the output shape depends on the input shape, but the _padding_ only depends on the kernel and dilation rate.)

The reason this matters is that using the input shape in the padding calculation means that we cannot convert networks with unknown input shapes, since the input shape is not known during the conversion process. If we use only the kernel & dilation, then we can convert those networks correctly.

This is important for us, as we are evaluating using NNVM/TVM in production. :)

**Comment by Huyuwei** [comment]
@thefiddler Thanks for pointing out. I will investigate it later. Could you please confirm that after the modification, the tests pass https://github.com/dmlc/nnvm/blob/master/tests/python/frontend/keras/test_forward.py

**Comment by thefiddler** [comment]
There are issues, investigating.

**Comment by Huyuwei** [comment]
@thefiddler I still hold the view that padding depends on input size, as documented in https://www.tensorflow.org/versions/r0.12/api_docs/python/nn/convolution.

Assume input = [32, 32], kernel=[3, 3], stride=[2, 2], according to your method, pad_top=1, pad_bottom=1; according to my method, pad_top=0, pad_bottom=1, a test will show the second is right.

After #322, pooling with same pad should pass tests with all input sizes, pool sizes, strides. Dilated conv should also be padded correctly, even though it's not supported by TVM backend.

**Comment by thefiddler** [comment]
After spending quite a bit more time, it seems that padding indeed does rely on the input size. I think.the difference is that tensorflow calculates padding dynamically at runtime, so it can deal with tensors of any shape. I am closing this in favour of #322
